apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/bytebuddy/CacheableTransformerDecorator.java

```diff
@@ -0,0 +1,195 @@
+                cacheDir = Files.createTempDirectory("class-cache").toFile();
+            } else {
+                cacheDir = new File(this.cacheDirBase + "/class-cache-" + RandomString.make());
+            }
+            cacheResolver = new FileCacheResolver(cacheDir);
+        } else {
+            cacheResolver = new MemoryCacheResolver();
+        }
+    }
+
+    @Override
+    public ResettableClassFileTransformer decorate(ResettableClassFileTransformer classFileTransformer) {
+        return new ResettableClassFileTransformer.WithDelegation(classFileTransformer) {
+
+            @Override
+            public byte[] transform(ClassLoader loader, String className, Class<?> classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) throws IllegalClassFormatException {
+                // load from cache
+                byte[] classCache = cacheResolver.getClassCache(loader, className);
+                if (classCache != null) {
+                    return classCache;
+                }
+
+                //transform class
+                classfileBuffer = classFileTransformer.transform(loader, className, classBeingRedefined, protectionDomain, classfileBuffer);
+
+                // save to cache
+                if (classfileBuffer != null) {
+                    cacheResolver.putClassCache(loader, className, classfileBuffer);
+                }
+
+                return classfileBuffer;
+            }
+        };
+    }
+
+    private static String getClassLoaderHash(ClassLoader loader) {
+        String classloader;
+        if (loader != null) {
+            classloader = Integer.toHexString(loader.hashCode());
+        } else {
+            //classloader is null for BootstrapClassLoader
+            classloader = "00000000";
+        }
+        return classloader;
+    }
+
+    interface ClassCacheResolver {
+
+        byte[] getClassCache(ClassLoader loader, String className);
+
+        void putClassCache(ClassLoader loader, String className, byte[] classfileBuffer);
+    }
+
+    static class MemoryCacheResolver implements ClassCacheResolver {
+        // classloaderHashcode@className -> class bytes
+        private Map<String, byte[]> classCacheMap = new ConcurrentHashMap<String, byte[]>();
+
+        @Override
+        public byte[] getClassCache(ClassLoader loader, String className) {
+            String cacheKey = getCacheKey(loader, className);
+            return classCacheMap.get(cacheKey);
+        }
+
+        @Override
+        public void putClassCache(ClassLoader loader, String className, byte[] classfileBuffer) {
+            String cacheKey = getCacheKey(loader, className);
+            classCacheMap.put(cacheKey, classfileBuffer);
+        }
+
+        private String getCacheKey(ClassLoader loader, String className) {
+            return getClassLoaderHash(loader) + "@" + className;
+        }
+    }
+
+    static class FileCacheResolver implements ClassCacheResolver {
```

Comment:
File cache is a very sensitive and tricky thing, thinking in this way, once 2+ agents are sharing the agent.config, and open this feature, most likely they will share this dir. What will happen if they active file cache?

In most cases, file cache is meaning you need a dir lock, and generate the dir automatically.

Reply:
class cache dir is a randomly generated random directory, different agents will isolate their respective class cache.
`cacheDir = new File(this.cacheDirBase + "/class-cache-" + RandomString.make());`

Reply:
Yes, but can't find the API using `public CacheableTransformerDecorator(ClassCacheMode cacheMode, String cacheDirBase) `. Do I miss anything?

The cacheDirBase seems never be used.

Reply:
So, base on your codes, it uses the temp folder only.

Reply:
From my understanding, I think the folder should be in the agent folder, in order to make sure writable. Also, should be more friendly to docker and k8s users. They would expect log folder is writable and will be new files added.

Reply:
> So, base on your codes, it uses the temp folder only.

Yes,  it uses the temp folder starts with 'class-cache' and with random suffix,  like 'class-cache23123131'

Reply:
> From my understanding, I think the folder should be in the agent folder, in order to make sure writable. Also, should be more friendly to docker and k8s users. They would expect log folder is writable and will be new files added.

The agent folder is writable?  Please give a sample path of class cache.

Reply:
`AgentPackagePath#getPath`. WriterFactory(logging) is using this to write.